app/src/pages/settings/CustomProviderFormComponents.tsx

```diff
@@ -0,0 +1,317 @@
+/**
+ * Reusable form field components for provider configuration
+ * These provide consistent styling and validation for advanced settings
+ */
+
+import { useCallback, useEffect, useRef, useState } from "react";
+import { Control, Controller, FieldPath } from "react-hook-form";
+import { JSONSchema7 } from "json-schema";
+
+import { Card, Flex, Text, View } from "@phoenix/components";
+import { CodeEditorFieldWrapper } from "@phoenix/components/code";
+import { JSONEditor } from "@phoenix/components/code/JSONEditor";
+import {
+  httpHeadersJSONSchema,
+  stringToHttpHeadersSchema,
+} from "@phoenix/schemas/httpHeadersSchema";
+
+import { ProviderFormData } from "./CustomProviderForm";
+
+/**
+ * Type-safe field path for provider form data
+ */
+type ProviderFormFieldPath = FieldPath<ProviderFormData>;
+
+/**
+ * Simple JSON validator for form UI - returns string | true for react-hook-form compatibility
+ * For more detailed validation, see validateJsonField in customProviderFormUtils
+ */
+function validateJsonForUI(value: string | undefined): string | true {
+  if (!value || value.trim() === "" || value.trim() === "{}") {
+    return true; // Empty is valid
+  }
+  try {
+    JSON.parse(value);
+    return true;
+  } catch {
+    return "Invalid JSON format";
+  }
+}
+
+interface AdvancedSettingsCardProps {
+  children: React.ReactNode;
+  title?: string;
+  description?: string;
+}
+
+export function AdvancedSettingsCard({
+  children,
+  title = "Advanced Settings",
+  description = "Optional advanced configuration. Leave blank to use sensible defaults.",
+}: AdvancedSettingsCardProps) {
+  return (
+    <div
+      onClick={(e) => {
+        // Only stop propagation if the click is on the card header/button
+        // Allow clicks inside the card body to propagate normally
+        if (
+          (e.target as HTMLElement).closest(".card__collapsible-button") ||
+          (e.target as HTMLElement).closest("header")
+        ) {
+          e.stopPropagation();
+        }
+      }}
+      style={{ width: "100%" }}
+    >
+      <Card
+        title={title}
+        collapsible
+        defaultOpen={false}
+        titleSeparator={false}
+      >
+        <View padding="size-200" paddingTop="size-100">
+          <Flex direction="column" gap="size-100">
+            <Text size="S" color="text-700">
+              {description}
+            </Text>
+            {children}
+          </Flex>
+        </View>
+      </Card>
+    </div>
+  );
+}
+
+interface JSONFieldProps {
+  name: ProviderFormFieldPath;
+  control: Control<ProviderFormData>;
+  label: string;
+  placeholder?: string;
+  description: string;
+  jsonSchema?: JSONSchema7;
+  validator?: (value: string) => string | true;
+}
+
+/**
+ * Internal editor component that handles state and validation
+ * Includes proper cleanup to prevent state updates after unmount
+ */
+function JSONEditorWithValidation({
```

Comment:
this component feels generic, I suppose even so far as to say I'm surprised it's not?